### PR TITLE
feat(api): AlertProfiles v2 service

### PR DESCRIPTION
**Pull request by dmurray-lacework**
## Summary

New Alert Profiles service added to go-sdk .

## How did you test this change?

Run `_examples/alert_profiles/main.go`
Unit Test `api/alert_profiles_test.go`

## Issue

https://lacework.atlassian.net/browse/ALLY-979

### CODE REVIEW

File: api/alert_profiles.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// AlertProfilesService is the service that interacts with
+// the AlertProfiles schema from the Lacework APIv2 Server
+type AlertProfilesService struct {
+	client    *Client
+	Templates AlertTemplatesService
```
Comment (dmurray-lacework) [inline api/alert_profiles.go:31]: AlertTemplates does not have a GET so cannot implement with our V2Service interface.

The api path is nested under AlertProfiles.
/api/v2/AlertProfiles/{alertProfileId}/AlertTemplates

File: api/alert_profiles.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+type AlertProfileField struct {
+	Name string `json:"name"`
+}
+
+type AlertProfileDescriptionKeys struct {
+	Name string `json:"name"`
+	Spec string `json:"spec"`
+}
+
+type AlertProfileAlert struct {
```
Comment (hazedav) [inline api/alert_profiles.go:137]: Isn't an `AlertProfileAlert` an `AlertTemplate` ?

File: api/alert_templates.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+func (svc *AlertTemplatesService) Create(alertProfileID string, template AlertProfileAlert) (
+	response AlertProfileResponse,
+	err error,
+) {
+	apiPath := fmt.Sprintf(apiV2AlertTemplates, alertProfileID)
+	err = svc.client.RequestEncoderDecoder("POST", apiPath, template, &response)
+	return
+}
+
+func (svc *AlertTemplatesService) Update(alertProfileID string, alertTemplateID string, template AlertProfileUpdate) (
```
Comment (hazedav) [inline api/alert_templates.go:36]: It appears the API lets you update an alert template in multiple ways?

You can either PATCH the `alerts` object via `v2/AlertProfiles/{alertProfileId}` or you can patch the actual alert template via `v2/AlertProfiles/{alertProfileId}/AlertTemplates/{alertTemplateName}`.

Given that the `AlertProfileService.Update` takes `AlertProfileUpdate`, shouldn't this be taking `AlertTemplateUpdate` ?

File: api/alert_profiles.go (hunk 2 above)
Comment (dmurray-lacework) [reply api/alert_profiles.go:137]: Good point, I'm leaning towards AlertTemplate

File: api/alert_templates.go (hunk 3 above)
Comment (dmurray-lacework) [reply api/alert_templates.go:36]: 👍  Yeh it should be AlertTemplateUpdate. Will fix

File: api/alert_profiles.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+// Basic usage: Initialize a new AlertProfileConfig struct, then
+//              use the new instance to do CRUD operations
+//
+//   client, err := api.NewClient("account")
+//   if err != nil {
+//     return err
+//   }
+//
+//   alertProfile := api.NewAlertProfile(
+//		"Foo",
```
Comment (afiune) [inline api/alert_profiles.go:45]: There is one string missing in this example, the `extends` string.

File: api/alert_profiles.go (hunk 2 above)
Comment (afiune) [reply api/alert_profiles.go:137]: yeah, this should be AlertTemplate 👍🏽

File: api/alert_profiles.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+type AlertProfileAlert struct {
+	Name        string `json:"name"`
+	EventName   string `json:"eventName"`
+	Description string `json:"description"`
+	Subject     string `json:"subject"`
+}
+
+type AlertProfileUpdate struct {
```
Comment (afiune) [inline api/alert_profiles.go:144]: I can't believe that we can't update the Alert Profile itself, this is only updating the alert templates 😕

File: api/alert_templates.go (hunk 3 above)
Comment (afiune) [reply api/alert_templates.go:36]: I am so confused as well @hazedav 😖

File: api/alert_templates.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	apiPath := fmt.Sprintf(apiV2AlertTemplatesFromGUID, alertProfileID, alertTemplateID)
+	err = svc.client.RequestEncoderDecoder("POST", apiPath, nil, nil)
+	return
+}
+
+type AlertTemplateUpdate struct {
+	EventName   string `json:"eventName"`
+	Description string `json:"description"`
+	Subject     string `json:"subject"`
+}
```
Comment (afiune) [inline api/alert_templates.go:57]: Should this have `omitempty`? or they all MUST be provided?

File: api/v2.go [hunk 7]
```diff
@@ -53,6 +54,7 @@ func NewV2Endpoints(c *Client) *V2Endpoints {
 	v2 := &V2Endpoints{c,
 		&UserProfileService{c},
 		&AlertChannelsService{c},
+		&AlertProfilesService{c, AlertTemplatesService{c}},
```
Comment (afiune) [inline api/v2.go:57]: shouldn't we follow this pattern for nested services? https://github.com/lacework/go-sdk/blob/main/api/v2_vulnerabilities.go#L23-L36

**Review by afiune — COMMENTED** [review]
Almost done, great comments about Alert Templates.

File: api/alert_templates.go (hunk 6 above)
Comment (dmurray-lacework) [reply api/alert_templates.go:57]: 👍

File: api/v2.go (hunk 7 above)
Comment (dmurray-lacework) [reply api/v2.go:57]: Is this not slightly different in the vulnerability services are at the same level. Templates is nested under profiles.

V2.Vulnerabilities.Hosts.Func() & V2.Vulnerabilities.Containers.Func()

vs 

V2.Vulnerabilities.AlertProfiles.Func()
V2.Vulnerabilities.AlertProfiles.Templates.Func(alertProfileID)

Comment (dmurray-lacework) [reply api/v2.go:57]: I'll refactor as you've suggested and see.

V2.Vulnerabilities.Alert.Profiles.Func()
V2.Vulnerabilities.Alert.Templates.Func()

Not sure if it matters but we also V2.Vulnerabilities.AlertRules and V2.Vulnerabilities.AlertChannels then not nested under "Alert"

Comment (dmurray-lacework) [reply api/v2.go:57]: Seeing it, I do prefer the suggested way 👍

File: api/alert_profiles.go (hunk 5 above)
Comment (dmurray-lacework) [reply api/alert_profiles.go]: Yes, the api only allows updating the alert templates
https://your-account.lacework.net/api/v2/docs#tag/AlertProfiles/paths/~1api~1v2~1AlertProfiles~1{id}/patch

**Review by hazedav — APPROVED** [review]
LGTM

**Pull request merged**